GetIntoTeachingApi/Controllers/TeachingEventsController.cs

```diff
@@ -163,6 +163,37 @@ public async Task<IActionResult> Get([FromRoute, SwaggerParameter("The `readable
             return Ok(new TeachingEventAddAttendee(candidate));
         }
 
+        [HttpPost]
+        [Route("")]
+        [SwaggerOperation(
+            Summary = "Adds or updates a teaching event.",
+            Description = "If the `id` is specified then the existing teaching event will be " +
+                          "updated, otherwise a new teaching event will be created.",
+            OperationId = "AddOrUpdateTeachingEvent",
+            Tags = new[] { "Teaching Events" })]
+        [ProducesResponseType(typeof(TeachingEvent), StatusCodes.Status201Created)]
+        [ProducesResponseType(typeof(IDictionary<string, string>), StatusCodes.Status400BadRequest)]
+        public async Task<IActionResult> AddOrUpdateTeachingEventAsync([FromBody] TeachingEvent teachingEvent)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            if (teachingEvent.Building != null)
+            {
+                _crm.Save(teachingEvent.Building);
+                await _store.SaveAsync(teachingEvent.Building);
+            }
+
+            _crm.Save(teachingEvent);

```

Comment:
The teaching event isn't persisted to the store here, just the building? I think the building will be persisted as part of the event if you do something like (also adds a comment as to why we persist to the store because I don't think its necessarily obvious):

```
if (teachingEvent.Building != null)
{
  _crm.Save(teachingEvent.Building);
}

_crm.Save(teachingEvent);
// Make the event/building immediately available in the cache.
_store.SaveAsync(teachingEvent);
```